### PR TITLE
[TEP074] Remove Image pipelineResources

### PR DESCRIPTION
**Pull request by JeromeJu**
# Changes
This commit removes the Image Resources support.

This PR removes `github.com/tektoncd/pipeline/pkg/apis/resource/v1alpha1/image`
the imagedigestexporter and the docs for image resources.It also removes the
 integration tests `helm_task_test` and`hermetic_taskrun_test` as these tests 
are testing building against image resources.

Removal of `pipelineResources`, as in https://github.com/tektoncd/pipeline/issues/5967 has been broken up into removal of each resources according to small PRs.

Part of #5967 
/kind misc

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
action required: please migrate off of `image` `pipelineresources` as it is removed, please refer to the doc at https://github.com/tektoncd/pipeline/blob/main/docs/pipelineresources.md#replacing-an-image-resource
`imagedigestexporter` image will not be built nor maintained from now on
```

### CODE REVIEW

File: docs/resources.md
Comment (lbernick) [inline docs/resources.md:587]:
```suggestion
Support for image resources has been removed. Please refer to the doc at https://github.com/tektoncd/pipeline/blob/main/docs/pipelineresources.md#replacing-an-image-resource for migration.
```

suggesting since "deprecation" implies that it still works, while "removal" implies that it does not.

**Review by lbernick — APPROVED** [review]

**Review by vdemeester — APPROVED** [review]

File: test/yamls/v1beta1/pipelineruns/pipelinerun.yaml
Comment (JeromeJu) [inline test/yamls/v1beta1/pipelineruns/pipelinerun.yaml:92]: Tried deleting these steps, but I have been wondering if this would make the whole test case meaningless without the image digest?

**Comment by JeromeJu** [comment]
[TestHelmDeployPipelineRun](https://github.com/tektoncd/pipeline/blob/main/test/helm_task_test.go#L42) and [TestKanikoTaskRun](https://github.com/tektoncd/pipeline/blob/main/test/kaniko_task_test.go#L44) are highly dependent on `imagedigestexporter`, I am trying to figure out in which way shall we refactor these tests?

**Comment by lbernick** [comment]
> [TestHelmDeployPipelineRun](https://github.com/tektoncd/pipeline/blob/main/test/helm_task_test.go#L42) and [TestKanikoTaskRun](https://github.com/tektoncd/pipeline/blob/main/test/kaniko_task_test.go#L44) are highly dependent on `imagedigestexporter`, I am trying to figure out in which way shall we refactor these tests?

Can you replace the PipelineResources in these test cases with Tasks, following the migration guide?

**Comment by lbernick** [comment]
> > [TestHelmDeployPipelineRun](https://github.com/tektoncd/pipeline/blob/main/test/helm_task_test.go#L42) and [TestKanikoTaskRun](https://github.com/tektoncd/pipeline/blob/main/test/kaniko_task_test.go#L44) are highly dependent on `imagedigestexporter`, I am trying to figure out in which way shall we refactor these tests?
> 
> Can you replace the PipelineResources in these test cases with Tasks, following the migration guide?

I want to change my feedback here-- I don't think these tests actually have value without the image pipelineresource, so you can go ahead and remove them.

**Comment by JeromeJu** [comment]
In this PR, one question regarding removing the image resource at https://github.com/tektoncd/pipeline/pull/6002 is that dependencies is within the apis at https://github.com/tektoncd/pipeline/pull/6002. But the image resources api itself is being removed in this PR. Then in this case, there are 2 options to move forward:
- etracting the image resources logics to /pkg/image to keep the imagedigestexporter image without image resources apis
- remove the imageDigestExporter with this removal of image resources

**Comment by dibyom** [comment]
Isn't imagedigestexporter part of the image resource?

**Comment by JeromeJu** [comment]
> Isn't imagedigestexporter part of the image resource?

I think so, so according to the dependencies we shall remove it along with the removal of the image `pipelineResources` apis?

**Review by vdemeester — APPROVED** [review]

File: examples/v1beta1/pipelineruns/pipelinerun.yaml
Comment (lbernick) [inline examples/v1beta1/pipelineruns/pipelinerun.yaml:93]: Can this example be updated to not use the imagedigestexporter image? I believe we mentioned in the WG yesterday that there's a suitable kaniko flag that can replace it.

Comment (dibyom) [reply examples/v1beta1/pipelineruns/pipelinerun.yaml:93]: yeah we should just replace this kaniko task with the latest version of kaniko from the Hub

Comment (JeromeJu) [reply examples/v1beta1/pipelineruns/pipelinerun.yaml:93]: > yeah we should just replace this kaniko task with the latest version of kaniko from the Hub
For v1beta1 I think we can but resolvers are behind beta feature flag in v1.

Comment (lbernick) [reply examples/v1beta1/pipelineruns/pipelinerun.yaml:93]: Thanks for opening the issue! even if we can't use remote resolution here we should still migrate off usage of this image though

Comment (JeromeJu) [reply examples/v1beta1/pipelineruns/pipelinerun.yaml]: Following up on the migration off the dependencies of the imageDigestExporter image, @QuanZhang-William has helped me with the investigation of errors with the latest knaiko task at https://github.com/tektoncd/pipeline/issues/5727#issuecomment-1306169832. According to the findings that the cataglog latest kaniko task works fine, it is the old `url` provided at https://github.com/GoogleContainerTools/skaffold/blob/main/examples/microservices/leeroy-web/Dockerfile was not compatible with the latest kaniko task. So here are two options moving forward to unblock this:

- [ ] delete the population of the image digest if the purpose of the `pipelinerun.yaml` was not to deploy the microservice from skaffold
- [ ] not to use the skaffold images examples for params, which would change the microservices to be built in `pipelinerun.yaml` also

Comment (JeromeJu) [reply examples/v1beta1/pipelineruns/pipelinerun.yaml]: Do we have any preference on maintaining the existent pipelinerun.yaml examples?

Comment (QuanZhang-William) [reply examples/v1beta1/pipelineruns/pipelinerun.yaml]: It looks to me that the goal of the example is to showcase a typical CICD pipeline (i.e. clone -> test -> build & push -> deploy). If that is the case I think we should go for #2.

File: test/yamls/v1beta1/pipelineruns/pipelinerun.yaml
Comment (pritidesai) [reply test/yamls/v1beta1/pipelineruns/pipelinerun.yaml:92]: I reviewed the examples and testes that are deleted so far as part of this PR. I would have preferred to replace the `image` resource with something equivalent but I am not sure what is that equivalent. 😞

File: examples/v1beta1/pipelineruns/pipelinerun.yaml
Comment (pritidesai) [reply examples/v1beta1/pipelineruns/pipelinerun.yaml]: `pipelinerun.yaml` is one of the original examples and has been a great resource showcasing a basic CI/CD use case, please preserve it if possible, thanks!

**Comment by JeromeJu** [comment]
/hold
This shall target to be the very last type of the removal of piplineResources according to https://github.com/tektoncd/pipeline/issues/5967#issuecomment-1428252386.

Blockers:
- [Migrate Kaniko Task off ImageDigestExporter](https://github.com/tektoncd/pipeline/pull/6094)
- [[TEP074] Remove Storage and Git PipelineResources](https://github.com/tektoncd/pipeline/pull/6150)

**Comment by afrittoli** [comment]
> Blockers:

> https://github.com/tektoncd/pipeline/pull/6094
> https://github.com/tektoncd/pipeline/pull/6150

Thanks @JeromeJu, it sounds like those two should be part of the milestone too, or this should be removed. wdyt?

**Comment by JeromeJu** [comment]
> > Blockers:
> 
> > #6094
> > #6150
> 
> Thanks @JeromeJu, it sounds like those two should be part of the milestone too, or this should be removed. wdyt?

Yeah it'd be appreciated if they could be added to the milestone @afrittoli  😺 . And sorry haven't been able to do so as there has been changes on the sequence of how the implementations should be planned out.

**Comment by JeromeJu** [comment]
/unhold
Not that #6150 is awaiting review, I think we could get this removed first.

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution_test.go
Comment (lbernick) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution_test.go:2510]: why remove this?

File: pkg/reconciler/taskrun/taskrun_test.go
Comment (lbernick) [inline pkg/reconciler/taskrun/taskrun_test.go:938]: `taskRunSubstitution` hasn't been deleted; I think this test case should remain

File: test/ko_test.go
Comment (lbernick) [inline test/ko_test.go:2]: why remove this file?

File: examples/v1beta1/taskruns/optional-resources-with-clustertask.yaml
Comment (lbernick) [inline examples/v1beta1/taskruns/optional-resources-with-clustertask.yaml:17]: I'd prefer just removing image resources from the examples that don't rely on building images rather than removing the examples entirely.

Comment (JeromeJu) [reply examples/v1beta1/taskruns/optional-resources-with-clustertask.yaml:17]: Thanks I will restore them.
Previously I saw them as cases that would only be meaningful if the outputs resources could be used as a digest. But that makes sense to leave the rest of git resources until removal of all resources.

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution_test.go
Comment (JeromeJu) [reply pkg/reconciler/pipelinerun/resources/pipelinerunresolution_test.go:2510]: In this case the image-resource was the missing one but now since it is being removed, we would have to change to the other resource that should be missing but required by the pipeline resource declaration.

File: pkg/reconciler/taskrun/taskrun_test.go
Comment (JeromeJu) [reply pkg/reconciler/taskrun/taskrun_test.go:938]: I think for this case, the expected steps and volumes will be (almost 🤔  )completely changed without imageResourceType as the outputs resources. Would it make sense to delete the `taskRunSubstitution` also since they are only referenced here?

File: test/ko_test.go
Comment (JeromeJu) [reply test/ko_test.go:2]: These were just the helper functions moved here earlier (they should have been deleted in previous PR also.)

File: pkg/reconciler/taskrun/taskrun_test.go
Comment (lbernick) [reply pkg/reconciler/taskrun/taskrun_test.go:938]: It looks like it gets some volumes from the image resource and some from the git resource. Can you just remove the ones that it gets from the image resource? (and same for params)

Comment (JeromeJu) [reply pkg/reconciler/taskrun/taskrun_test.go]: That makes sense for maintaining code coverage:) Thanks!

**Pull request merged**
